### DISCUSSION TITLE
How to make each level automatically align with it

### DISCUSSION THREAD

**Discussion by netforfreedom** [freeplane/freeplane#1702] [Ideas]
version:v1.11.2 
The content of each sub node is different, resulting in different lengths of nodes. We hope to provide an explanation of its functionality,

Ability to automatically align nodes at each level. This makes the map more hierarchical and easier to understand at a glance.
Please refer to the picture, which shows the effect before it:
![1](https://github.com/freeplane/freeplane/assets/39705624/c0014c1c-049f-48b7-89a6-9cfbf6dcdec9)

------
This is for the subsequent effect
![2](https://github.com/freeplane/freeplane/assets/39705624/7698bb37-e725-40b4-b48b-9753a2638bd6)

**Comment by metbril** [comment]
This would be nice.

current workaround is to set min and max width for the parent node.

**Comment by euu2021** [comment]
I think it's possible to create a script that does that. It would be a calculation like this: depending on the width of the parent node, a certain horizontal shift is added to the child node, using `node.horizontalShift`.

The public API has no method to retrieve the width of the node, but [this thread](https://sourceforge.net/p/freeplane/discussion/758437/thread/f5517e8b/) explains how to retrieve the width using internal methods:
```
import org.freeplane.view.swing.map.NodeView
def nodeView = node.delegate.viewers.find() {it instanceof NodeView}
c.statusInfo = "width is ${nodeView?.mainView?.width}"
```

Tomorrow, I will try to create the script.

**Reply by euu2021** [reply]
I did some tests and found that the width of the node changes according to the zoom level, so this script is not possible.

**Comment by dpolivaev** [comment]
I converted the issue to discussion. I think it is doable but differently: the node width would be changed. However, it would add one more option and sometimes I think there are already too many options.

Please comment.

**Reply by euu2021** [reply]
> I think it is doable but differently: the node width would be changed.

Sorry, I don't understand the idea. Does the idea involve changing the node width?

**Reply by dpolivaev** [reply]
yes

**Reply by dpolivaev** [reply]
instead of changing the distance to the parent node, I would rather change the parent node width so that all siblings have the same width

**Reply by euu2021** [reply]
> instead of changing the distance to the parent node, I would rather change the parent node width so that all siblings have the same width

If I understand correctly, this would force the parents and all the siblings of the parent to be the same width, is it? This makes me remember this script, from the scripts collection: https://docs.freeplane.org/scripting/Scripts_collection.html#toggle-block-view-for-all-siblings

Anyway, I don't see why it would deserve a new vanilla option. I always set equal min and max width for all nodes, and I think it is enough for most usecases. More specific usecases can be solved with scripts.

**Reply by dpolivaev** [reply]
The difference is that layout it triggered automatically by any change in the map and the script should be called by the user explicitly.

**Reply by euu2021** [reply]
I see.

I just note that the example shown by the OP has the nodes keeping their original widths. So, it's not clear if there is anyone interested in the proposed feature. I'm not. OP needs to speak for himself.

**Comment by EdoFro** [comment]
an easy way to have equally wide nodes is to use automatic level styling and define for each level style that its min and max width are equal.

**Reply by rubcap** [reply]
I guess this works until you drag one of your first- or N-level nodes to another position? Then, it becomes really cumbersome to align them again. See my suggestion in this thread for an alignment feature.

**Reply by rubcap** [reply]
I stand corrected ... see below.

**Comment by rubcap** [comment]
Hi all,

First of all, thank you for the amazing software that Freeplane is ❤️.

A similar idea would be to have Alignment options in the Formatting menu. This would allow me to select nodes and align them on the horizontal (or vertical) start/center/end position of the first selected reference node.

I've made this quick diagram to illustrate the concept.
![Feedback_Freeplane_alignment_nodes](https://github.com/freeplane/freeplane/assets/26146350/bbded889-bbb5-4b7e-be60-8dd063eab23c)

**Reply by euu2021** [reply]
Freeplane already has something like this. Make a small movement in the node to be the reference for the alignment, and the others will align:
![java_cWnfZRYf2e](https://github.com/freeplane/freeplane/assets/77707706/db956635-010d-484c-98fe-fe03e768f5fe)

**Reply by rubcap** [reply]
No way, that's amazing! Thank you so much for letting me know.
